tokenbucket.go

```diff
@@ -465,3 +468,83 @@ func (t *TokenBucketRedis) SetState(ctx context.Context, state TokenBucketState)
 
 	return errors.Wrap(err, "failed to save keys to redis")
 }
+
+// TokenBucketDynamoDB is a DyanamoDB implementation of a TokenBucketStateBackend.
+type TokenBucketDynamoDB struct {
+	leakyBucket *LeakyBucketDynamoDB
+}
+
+// NewTokenBucketDynamoDB creates a new TokenBucketDynamoDB instance.
+// PartitionKey is the key used to store all the this implementation in DynamoDB.
+//
+// TableProps describe the table that this backend should work with. This backend requires the following on the table:
+// * TTL
+//
+// TTL is the TTL of the stored item.
+//
+// If raceCheck is true and the item in DynamoDB are modified in between State() and SetState() calls then
+// ErrRaceCondition is returned.
+func NewTokenBucketDynamoDB(client *dynamodb.Client, partitionKey string, tableProps DynamoDBTableProperties, ttl time.Duration, raceCheck bool) *TokenBucketDynamoDB {
+	return &TokenBucketDynamoDB{
+		leakyBucket: NewLeakyBucketDynamoDB(client, partitionKey, tableProps, ttl, raceCheck),
+	}
+}
+
+// State gets the bucket's state from DynamoDB.
+func (t *TokenBucketDynamoDB) State(ctx context.Context) (TokenBucketState, error) {
+	resp, err := dynamoDBGetItem(ctx, t.leakyBucket.client, t.leakyBucket.getGetItemInput())
+
+	if err != nil {
+		return TokenBucketState{}, err
+	}
+
+	return t.loadStateFromDynamoDB(resp)
+}
+
+// SetState updates the state in DynamoDB.
+func (t *TokenBucketDynamoDB) SetState(ctx context.Context, state TokenBucketState) error {
+	input := t.getPutItemInputFromState(state)
+
+	var err error
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		_, err = dynamoDBputItem(ctx, t.leakyBucket.client, input)
+	}()
+
+	select {
+	case <-done:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+
+	return err
+}
+
+const dynamoDBBucketAvailableKey = "Available"
+
+func (t TokenBucketDynamoDB) getPutItemInputFromState(state TokenBucketState) *dynamodb.PutItemInput {
```

Comment:
Struct TokenBucketDynamoDB has methods on both value and pointer receivers. Such usage is not recommended by the Go Documentation.